Validation/HGCalValidation/test/HGCalWaferHitCheck.cc

```diff
@@ -0,0 +1,263 @@
+    edm::Handle<HGCalDigiCollection> theDigiContainer;
+    iEvent.getByToken(digiHitSource_, theDigiContainer);
+    if (theDigiContainer.isValid()) {
+      if (verbosity_ > 0)
+        edm::LogVerbatim("HGCalValidation") << nameDetector_ << " with " << theDigiContainer->size() << " Digis";
+      analyzeHits(nameDetector_, *(theDigiContainer.product()));
+    } else if (verbosity_ > 0) {
+      edm::LogVerbatim("HGCalValidation") << "DigiContainer does not exist for " << nameDetector_;
+    }
+  } else {
+    edm::Handle<HGCRecHitCollection> theRecHitContainer;
+    iEvent.getByToken(digiHitSource_, theRecHitContainer);
+    if (theRecHitContainer.isValid()) {
+      if (verbosity_ > 0)
+        edm::LogVerbatim("HGCalValidation") << nameDetector_ << " with " << theRecHitContainer->size() << " hits";
+      analyzeHits(nameDetector_, *(theRecHitContainer.product()));
+    } else if (verbosity_ > 0) {
+      edm::LogVerbatim("HGCalValidation") << "RecHitContainer does not exist for " << nameDetector_;
+    }
+  }
+}
+
+template <class T>
+void HGCalWaferHitCheck::analyzeHits(std::string const& name, T const& hits) {
+  for (auto const& hit : hits) {
+    uint32_t id = hit.id();
+    int zside, type, layer, waferU, waferV;
+    if (ifNose_) {
+      HFNoseDetId detId = HFNoseDetId(id);
+      waferU = detId.waferU();
+      waferV = detId.waferV();
+      type = detId.type();
+      layer = detId.layer();
+      zside = detId.zside();
+    } else {
+      HGCSiliconDetId detId = HGCSiliconDetId(id);
+      waferU = detId.waferU();
+      waferV = detId.waferV();
+      type = detId.type();
+      layer = detId.layer();
+      zside = detId.zside();
+    }
+    int index = HGCalWaferIndex::waferIndex(layer, waferU, waferV, false);
+    int typef = hgcons_->waferType(layer, waferU, waferV, true);
+    if (zside < 0)
+      index *= -1;
+    auto itr = occupancy_.find(index);
+    if (itr == occupancy_.end())
+      occupancy_[index] = 1;
+    else
+      ++occupancy_[index];
+    if (type != typef) {
+      auto ktr = badTypes_.find(index);
+      if (ktr == badTypes_.end())
+        badTypes_[index] = 1;
+      else
+        ++badTypes_[index];
+      if (verbosity_ == 1)
+        edm::LogVerbatim("HGCalValidation")
+            << "Detector " << name << " zside = " << zside << " layer = " << layer << " type = " << type << ":" << typef
+            << " wafer = " << waferU << ":" << waferV << " index " << index;
+    }
+    if (verbosity_ > 1)
+      edm::LogVerbatim("HGCalValidation")
+          << "Detector " << name << " zside = " << zside << " layer = " << layer << " type = " << type << ":" << typef
+          << " wafer = " << waferU << ":" << waferV;
+  }
+}
+
+// ------------ method called when starting to processes a run  ------------
+void HGCalWaferHitCheck::beginRun(const edm::Run&, const edm::EventSetup& iSetup) {
+  edm::ESHandle<HGCalDDDConstants> pHGDC;
+  iSetup.get<IdealGeometryRecord>().get(nameDetector_, pHGDC);
```

Comment:
Please also declare the consumption EventSetup products and use ESGetToken to get them
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToGetDataFromES#In_ED_module